plugins/mana/plugin.go

```diff
@@ -371,8 +414,255 @@ func GetPendingMana(value float64, n time.Duration) float64 {
+		})
+		sort.Slice(logs[ID].Revoke, func(i, j int) bool {
+			return logs[ID].Revoke[i].Time.Before(logs[ID].Revoke[j].Time)
+		})
+	}
+
+	return logs, err
+}
+
+// GetPastConsensusManaVectorMetadata gets the past consensus mana vector metadata.
+func GetPastConsensusManaVectorMetadata() *mana.ConsensusBasePastManaVectorMetadata {
+	cachedObj := consensusBaseManaPastVectorMetadataStorage.Get([]byte(mana.ConsensusBaseManaPastVectorMetadataStorageKey))
+	cachedMetadata := &mana.CachedConsensusBasePastManaVectorMetadata{CachedObject: cachedObj}
+	defer cachedMetadata.Release()
+	return cachedMetadata.Unwrap()
+}
+
+// GetPastConsensusManaVector builds a consensus base mana vector in the past.
+func GetPastConsensusManaVector(t time.Time) (*mana.ConsensusBaseManaVector, []mana.Event, error) {
+	baseManaVector, err := mana.NewBaseManaVector(mana.ConsensusMana)
+	if err != nil {
+		return nil, nil, err
+	}
+	cbmvPast := baseManaVector.(*mana.ConsensusBaseManaVector)
+	cachedObj := consensusBaseManaPastVectorMetadataStorage.Get([]byte(mana.ConsensusBaseManaPastVectorMetadataStorageKey))
+	cachedMetadata := &mana.CachedConsensusBasePastManaVectorMetadata{CachedObject: cachedObj}
+	defer cachedMetadata.Release()
+
+	if cachedMetadata.Exists() {
+		metadata := cachedMetadata.Unwrap()
+		if t.After(metadata.Timestamp) {
+			consensusBaseManaPastVectorStorage.ForEach(func(key []byte, cachedObject objectstorage.CachedObject) bool {
+				cachedPbm := &mana.CachedPersistableBaseMana{CachedObject: cachedObject}
+				defer cachedPbm.Release()
+				p := cachedPbm.Unwrap()
+				err = cbmvPast.FromPersistable(p)
+				if err != nil {
+					log.Errorf("error while restoring %s mana vector from storage: %w", mana.ConsensusMana.String(), err)
+					baseManaVector, _ := mana.NewBaseManaVector(mana.ConsensusMana)
+					cbmvPast = baseManaVector.(*mana.ConsensusBaseManaVector)
+					return false
+				}
+				return true
+			})
+		}
+	}
+
+	var eventLogs mana.EventSlice
+	consensusEventsLogStorage.ForEach(func(key []byte, cachedObject objectstorage.CachedObject) bool {
+		cachedPe := &mana.CachedPersistableEvent{CachedObject: cachedObject}
+		defer cachedPe.Release()
+		pe := cachedPe.Unwrap()
+		if pe.Time.After(t) {
+			return true
+		}
+
+		// already consumed in stored base mana vector.
+		if cachedMetadata.Exists() && cbmvPast.Size() > 0 {
+			metadata := cachedMetadata.Unwrap()
+			if pe.Time.Before(metadata.Timestamp) {
+				return true
+			}
+		}
+
+		var ev mana.Event
+		ev, err = mana.FromPersistableEvent(pe)
+		if err != nil {
+			return false
+		}
+		eventLogs = append(eventLogs, ev)
+		return true
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	eventLogs.Sort()
+	err = cbmvPast.BuildPastBaseVector(eventLogs, t)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = cbmvPast.UpdateAll(t)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return cbmvPast, eventLogs, nil
+}
+
+func checkConsensusEventStorage() {
+	size := consensusEventsLogStorage.GetSize()
```

Comment:
@acha-bill another bug/feature I found while testing is that `objectStorage.GetSize()` returns the size of the cache, not the size of the elements in cache + persistence. The cache content is persisted after couple of seconds, so  `size` will never be bigger, than `maxConsensusEventsInStorage`. The actual pruning of the events and constructing the past vector never happens.

Reply:
Good catch. I'll load the size of the storage on start and increment it as events are being stored.
Sometime in the future we should have an `objectstorage.GetSize(skipCache ...bool)`

Reply:
submit a ticket in `hive` ;)